File: daily/2024/2458_Height_of_Binary_Tree_After_Subtree_Removal_Queries.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

TreeNode* create_tree(vector<int>& nodes){
    if (nodes.empty()) return nullptr;
    TreeNode* root = new TreeNode(nodes[0]);
    int i = 1;
    int n = nodes.size();
    queue<TreeNode*> que;
    que.push(root);
    while(!que.empty()){
        TreeNode* curr = que.front();
        que.pop();
        if (i < n && nodes[i] != -1){
            curr->left = new TreeNode(nodes[i]);
            que.push(curr->left);
        }
        i++;
        if (i < n && nodes[i] != -1){
            curr->right = new TreeNode(nodes[i]);
            que.push(curr->right);
        }
        i++;
    }
    return root;
}

class Solution {
public:
    int dfs(TreeNode* root, int curr_level, unordered_map<int,pair<int,int>>& m){
        // cout << "dfs" << root->val << endl;
        int under_max = 0;
        if (root->left){
            under_max = max(under_max, dfs(root->left, curr_level+1, m));
        }
        if (root->right){
            under_max = max(under_max, dfs(root->right, curr_level+1, m));
        }
        pair<int,int> curr_pair = {curr_level, under_max};
        m[root->val] = curr_pair;
        // cout << root->val << " : {" << curr_level << "," << under_max << "}" << endl;
        return under_max + 1;
    }
    void find_ans(TreeNode* node, int limit, unordered_map<int,pair<int,int>>& m, vector<int>& ans){
        int left = 0;
        if (node->left) left = max(left, m[node->left->val].first + m[node->left->val].second);
        int right = 0;
        if (node->right) right = max(right, m[node->right->val].first + m[node->right->val].second);
        if (node->left) find_ans(node->left, max(limit, right), m, ans);
        if (node->left) ans[node->left->val] = max({ans[node->left->val], limit, right});
        if (node->right) find_ans(node->right, max(limit, left), m, ans);
        if (node->right) ans[node->right->val] = max({ans[node->right->val],limit, left});
        ans[node->val] = max(ans[node->val], m[node->val].first - 1);
        return;
    }
    vector<int> treeQueries(TreeNode* root, vector<int>& queries) {
        // val : {level, under_max}
        unordered_map<int,pair<int,int>> m;
        dfs(root, 0, m);
        // for (auto i : m) cout << i.first << " : {" << i.second.first << "," << i.second.second << "}" << endl;
        int n = m.size();
        vector<int> ans(n+1);
        find_ans(root, 0, m, ans);
        // for (auto i : ans) cout << i << " ";
        int q_m = queries.size();
        vector<int> q(q_m);
        for (int i = 0; i < q_m; i++){
            q[i] = ans[queries[i]];
        }
        return q;
    }
};

int main(){
    vector<int> nodes = {1,3,4,2,-1,6,5,-1,-1,-1,-1,-1,7};
    vector<int> queries = {4};
    // vector<int> nodes = {5,8,9,2,1,3,7,4,6};
    // vector<int> queries = {3,2,4,8};
    // vector<int> nodes = {1,-1,5,3,-1,2,4};
    // vector<int> queries = {3,5,4,2,4};
    TreeNode* root = create_tree(nodes);
    Solution S;
    vector<int> ans = S.treeQueries(root, queries);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/2458_Height_of_Binary_Tree_After_Subtree_Removal_Queries.py
import os
import sys
from typing import Optional, List
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Tree:
    def __init__(self, nodes):
        if not nodes:
            self.root = None
        else:
            self.root = TreeNode(nodes[0])
            i = 1
            n = len(nodes)
            que = deque([self.root])
            while que and i < n:
                curr = que.popleft()
                if i < n and nodes[i] != -1:
                    curr.left = TreeNode(nodes[i])
                    que.append(curr.left)
                i += 1
                if i < n and nodes[i] != -1:
                    curr.right = TreeNode(nodes[i])
                    que.append(curr.right)
                i += 1

class Solution:
    def dfs_left_first(self, root: TreeNode, level: int):
        if not root:
            return
        self.left_first[root.val] = self.limit
        self.limit = max(level, self.limit)
        self.dfs_left_first(root.left, level+1)
        self.dfs_left_first(root.right, level+1)

    def dfs_right_first(self, root: TreeNode, level: int):
        if not root:
            return
        self.right_first[root.val] = self.limit
        self.limit = max(level, self.limit)
        self.dfs_right_first(root.right, level+1)
        self.dfs_right_first(root.left, level+1)

    def treeQueries(self, root: Optional[TreeNode], queries: List[int]) -> List[int]:
        self.left_first = {}
        self.right_first = {}
        self.limit = 0
        self.dfs_left_first(root, 0)
        self.limit = 0
        self.dfs_right_first(root, 0)
        ans = []
        for q in queries:
            ans.append(max(self.left_first.get(q, 0), self.right_first.get(q, 0)))
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nodes = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            queries = list(map(int, lines[1].split('=')[1].strip()[1:-2].split(',')))
            T = Tree(nodes)
            dataset.append((T.root, queries))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for root, queries in dataset:
        results.append(solution.treeQueries(root, queries))
    for index, result in enumerate(results):
        print(f'Example : {index + 1} : {result}')